### PR TITLE
Hold the multi-process byte ranges in the existing open modes

### PR DESCRIPTION
**Pull request by cberner**
First PR of the multi-process implementation, rebuilt on the revised design in `docs/design.md` (single file, byte-range locks). No new public API.

The multi-process locking protocol coordinates processes through byte-range locks on the database file. Per the design's "Single process" and "Read-only" modes, the existing open paths now hold the entire byte range: exclusively for a writable `Database`, shared for a `ReadOnlyDatabase`, released at close. A conflict on the range reports `DatabaseAlreadyOpen`, which is what will refuse an ordinary open while a future multi-process handle has the file open, and vice versa.

Details:
* On Linux the range locks are held in addition to the whole-file lock: `flock()` and `fcntl()` locks are separate namespaces there and cannot see each other. On the Apple platforms the two share one lock table and conflict with each other -- even through the same file description, which macOS CI demonstrated by refusing every open that tried to hold both -- so the range lock is taken alone there; the shared table is exactly what makes it conflict with everything else, including the whole-file locks of older redb versions. The design doc's mode sections are updated to state both cases.
* Open file description locks (`F_OFD_SETLK`), never POSIX record locks: a record lock is dropped when any descriptor for the file closes anywhere in the process, which a library embedded in an arbitrary program cannot rule out. This adds the `libc` dependency on Linux and the Apple platforms.
* The whole range is locked with `l_len = 0` ("to the largest possible offset"): an explicit length cannot reach the last addressable byte, because the offsets `fcntl` takes are signed.
* Windows needs no change: std's whole-file lock is a `LockFileEx` over every byte, so it already conflicts with any range the protocol will lock.
* Other platforms keep the whole-file lock alone; the multi-process modes will not be offered there. A filesystem that refuses range locks degrades the same way an unsupported whole-file lock always has: the database opens, with a logged warning.
* Unit tests probe the protocol's actual byte map (the header lock bytes, the coordination bytes at `2^62`, the active transaction range, and the last addressable byte) through a second file description, in both directions.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01SJFSfcturbVcnPqY5CNzQv

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -475,6 +475,368 @@ All pages referenced by a savepoint must be contained in the above, because it i
...
 
+# Multi-process access (prototype)
+
+`MultiProcessDatabase` extends the epoch based reclamation above across processes. Everything it
+needs from the operating system is ordinary file I/O and the whole-file advisory locks in
+`std::fs::File`; there is no shared memory and no `unsafe`.
+
+This section describes the design as a whole. It is built in steps, so a given release implements
+only part of what is written here; the type is behind the `experimental-multiprocess` feature until
+all of it is in place.
```
Comment (cberner) [inline docs/design.md]: Remove. design.md should describe the final step. It is not a working list of steps

File: docs/design.md [hunk 2]
```diff
@@ -475,6 +475,368 @@ All pages referenced by a savepoint must be contained in the above, because it i
...
+an exclusive advisory lock on the file itself, which is what stops a second process opening it --
+and which also stops any other process reading it.
+
+`data.redb` therefore carries no lock of its own: every process using the database has it open at
+once, so an exclusive lock is impossible. A shared one is no better, because on Windows a shared
+range denies writes to every process *including* the one holding the lock, so a writer cannot hold
+one on a file it writes to at all. All reader/writer coordination lives in the files beside it, and
+those are what keep the processes out of each other's way. Nothing about the file itself would stop
+a plain `Database` from opening it and writing underneath them, so `Database::open()` and its
+siblings refuse a `data.redb` with a marker beside it.
```
Comment (cberner) [inline docs/design.md]: shorten all of this. It's pretty obvious why we need a directory. just a few sentences will do

File: docs/design.md [hunk 3]
```diff
@@ -475,6 +475,368 @@ All pages referenced by a savepoint must be contained in the above, because it i
...
+
+`write.lock` is taken before anything else in the directory is opened, so that the process which
+gets it has the directory to itself, including while it is being created. It is held by the storage
+backend rather than beside the `Database`, because a live write transaction keeps the database open
+past the point where the handle is dropped: a lock released when the handle went away would let
+another process start writing while that transaction was still running, and tying it to the backend
+gives it exactly the lifetime of the open file.
```
Comment (cberner) [inline docs/design.md]: design.md should limit its use of implementation specific terms. So describe the protocol and not the exact structs involved -- so avoid saying things like "`Database`". Apply this in other areas too.

File: docs/design.md [hunk 4]
```diff
@@ -475,6 +475,368 @@ All pages referenced by a savepoint must be contained in the above, because it i
...
+
+A platform without file locking cannot support any of this safely, so opening fails there rather
+than warning and continuing as `Database` does.
+
+`registry.lock` holds the writer mode, the last transaction that has been made durable, the highest
+free horizon any writer has used, and a counter of how many times a horizon has been announced.
```
Comment (cberner) [inline docs/design.md]: Describe the exact layout

**Review by cberner — COMMENTED** [review]
I left some comments on the design doc. But didn't read further than my last comment. The rest is way too verbose. Tighten it up and match the style of the rest of the design doc

**Comment by cberner** [comment]
Rewrote the multi-process section of design.md: 362 lines -> 217. Pushed to all six branches (the doc lives in this PR; the rest just carry it forward unchanged).

- **"Remove. design.md should describe the final step."** Gone, along with `(prototype)` in the heading. This reverses the Codex suggestion above, which asked for the opposite -- yours wins.
- **"shorten all of this ... just a few sentences will do."** The directory rationale is one paragraph now: exclusion has to move off the data file, an exclusive lock is impossible because every process has the file open at once, and a shared lock is worse still on Windows.
- **"limit its use of implementation specific terms ... Apply this in other areas too."** Applied throughout the section: no `Database`, `MultiProcessDatabase`, `Durability::None`, `TransactionTracker`, `begin_write`, `create()`/`open()`, `std::fs`, or `DatabaseError` variants. It reads as "a writer", "a reader", "creating the directory", "the single-file format described above".
- **"Describe the exact layout."** `metadata` (12 bytes) and `registry.lock` (32 bytes) now have byte-offset bullet lists and per-field descriptions, in the same style as the super-header section above.
- **"way too verbose ... match the style of the rest of the design doc."** The two sections you hadn't reached -- `Recognizing a directory` and `What a call that fails leaves behind`, ~170 lines of create/open edge cases -- are now one 34-line `Creating a directory` section covering the ordering rules that are actually protocol. The case-by-case reasoning that came out is still in code comments, where it belongs. I also folded the multi-process `Assumptions` subsection into the existing `# Assumptions about underlying media` at the end of the doc rather than repeating the heading.

Only design.md changed; no code moved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/multiprocess_tests.rs [hunk 5]
```diff
@@ -0,0 +1,316 @@
+//! Tests for the multi-process database interface.
+//!
+//! At this step the interface allows only one process to have the database open, so what there is
+//! to test is the directory layout and the lock file that enforces that. The lock is exercised
+//! both from a second handle in this process -- file locks belong to the open file description
+//! rather than the process, so a second handle is excluded exactly as a second process is -- and
+//! from a real child process, which is what the lock is ultimately for.
+
+#![cfg(all(feature = "experimental-multiprocess", not(target_os = "wasi")))]
```
[the remark this answers is not in the document]
Comment (cberner) [reply tests/multiprocess_tests.rs]: Checked this one and I don't think it holds, on either half.

**The combination doesn't reach this gate.** `redb_no_std` requires `panic = "abort"`, enforced by a deliberate `compile_error!` in `lib.rs`:

```
#[cfg(all(redb_no_std, panic = "unwind"))]
compile_error!("redb without the standard library requires panic = \"abort\": ...");
```

The dev profile unwinds, so `--no-default-features --features experimental-api-5,experimental-multiprocess` on the host fails there first, with or without this gate, and with or without the multiprocess feature. That is why CI builds no_std only for `thumbv7em-none-eabihf`. Running CI's own command with the feature added:

```
cargo check -p redb --target thumbv7em-none-eabihf --no-default-features --features experimental-api-5                          # ok
cargo check -p redb --target thumbv7em-none-eabihf --no-default-features --features experimental-api-5,experimental-multiprocess # ok
```

Integration tests aren't built for a bare-metal target, so `tests/` never enters the supported no_std configuration.

**And the suggested fix isn't available here.** `redb_no_std` is set by redb's `build.rs` through `cargo:rustc-cfg`, which applies only to the crate being built -- the library. An integration test in `tests/` is a separate crate and never sees it, so `not(redb_no_std)` in this file would always be true and would change nothing. That is why the gate uses the feature and the target instead of copying `lib.rs`.

The alternative suggestion, having `experimental-multiprocess` enable `std`, would mean `--no-default-features --features experimental-multiprocess` silently turns `std` back on. That seems worse than the current arrangement, where asking for a feature that needs `std` in a no_std build simply leaves the type unexported.

Leaving the gate as it is.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/design.md [hunk 6]
```diff
@@ -475,6 +475,226 @@ All pages referenced by a savepoint must be contained in the above, because it i
...
+lot. That holds on Unix only: elsewhere there is no portable way to reach a directory handle or to
+make a rename write-through, so this is a known gap. A power loss just after creation returns can
+lose the marker while keeping the database, leaving a directory that is refused until it is created
+again. No data is lost.
+
+One state cannot be repaired automatically. A marked directory with no `data.redb` but a temporary
+beside it means the rename that would have promoted it did not survive -- and since the marker went
+in after that rename, the temporary is the database rather than an abandoned attempt. Nothing about
+the files distinguishes the two, so creation refuses and says what to rename. This is reachable only
+where a directory sync did not happen.
```
Comment (cberner) [inline docs/design.md]: This is way too long and pedantic. The design doc should focus on the interesting parts of redb, not the rules of fsync

**Review by cberner — COMMENTED** [review]
The design doc is still too verbose. And the intended audience is very advanced programmers, so there is no need to say obvious things like "That is what makes the scheme portable, since a lock is mandatory on some platforms and a file another process holds could not be read at all"

Trim it some more, and improve it.

**Comment by cberner** [comment]
Trimmed again: 219 -> 158 lines (362 before your first round). Pushed through all six branches; the stack is also rebased onto current master, which it had fallen behind.

**The fsync rules are gone.** The whole `Creating a directory` section went -- rename-into-place, the directory and parent syncs, the Unix-only gap, the one state that needs a manual rename. What survives is two sentences in the layout preamble, because they are the part the *protocol* rests on rather than durability bookkeeping: the files are renamed into place with `metadata` last, so a directory is recognized only once it holds a finished database, and a failed call leaves no marker. The rest lives in code comments. The only `fsync` left in the section is the extra one a 2-phase commit costs.

**The obvious explanations are gone**, including the sentence you quoted. Others in the same vein that came out: why a lock file another process holds cannot be read, why the shared lock on `txn/<id>` is shared rather than exclusive, why a reader can trust a published transaction id, what a `Cow`-style rename buys. Where a justification stayed, it is one an advanced reader still could not derive from the mechanism alone -- the cross-process visibility argument for forcing 2-phase commits, the `H > S + 1` staleness bound, why the horizon needs a sequence number beside it.

Section is now: layout table, the two byte layouts, writer modes, `txn/` pinning and the registry-lock atomicity, savepoints, header reload, cache invalidation, crash recovery.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/design.md [hunk 7]
```diff
@@ -499,3 +658,9 @@ Therefore, we make only a few assumptions about the guarantees provided by the u
    a range of bytes in a file, no bytes outside of that range will change,
    even if the write occurs just before a crash or power failure. sqlite makes this same
    assumption, by default, in all modern versions.
+
+Multi-process access assumes one thing more: that a read of the super-header cannot observe a
+partially completed concurrent write of it -- the read/write atomicity POSIX requires of regular
+files. Nothing else is ever read while it is being written, since every other page is
+copy-on-write, and the protocol keeps a page from being rewritten while another process can still
+reach it.
```
Comment (cberner) [inline docs/design.md]: Make sure to document this on the StorageBackend trait

Comment (cberner) [reply docs/design.md]: Done -- on the `StorageBackend` trait itself, since a backend author has no reason to read design.md's media assumptions:

```rust
/// A backend for a multi-process database must satisfy one requirement beyond the rest of this
/// trait: a [`read`](Self::read) of a range that another process is writing at the same time must
/// not observe a partially completed write of it. That applies only to the database header, the
/// sole range ever read while it is being written; every other page is copy-on-write, and the
/// protocol keeps one from being rewritten while another process can still reach it. Regular files
/// satisfy this, since POSIX requires read/write atomicity of them.
```

Prose rather than a link to `MultiProcessDatabase`, which would not resolve with the feature off. `cargo doc` is clean both with and without it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/db.rs [hunk 8]
```diff
@@ -41,6 +41,13 @@ use log::{debug, warn};
 /// Implements persistent storage for a database.
 ///
 /// Failures are reported as [`io::Error`], which is [`std::io::Error`] whenever std is available.
+///
+/// A backend for a multi-process database must satisfy one requirement beyond the rest of this
+/// trait: a [`read`](Self::read) of a range that another process is writing at the same time must
+/// not observe a partially completed write of it. That applies only to the database header, the
+/// sole range ever read while it is being written; every other page is copy-on-write, and the
+/// protocol keeps one from being rewritten while another process can still reach it. Regular files
+/// satisfy this, since POSIX requires read/write atomicity of them.
```
Comment (cberner) [inline src/db.rs]: Remove this

File: src/lib.rs [hunk 9]
```diff
@@ -115,6 +117,9 @@ pub mod io;
 mod io;
 #[cfg(feature = "experimental-api-5")]
 mod key_range;
+// Needs std::fs and std::path, like the rest of the file-backed API
+#[cfg(all(feature = "experimental-multiprocess", not(redb_no_std)))]
+mod multi_process;
```
Comment (cberner) [inline src/lib.rs]: Let's use "multiprocess" consistently for all the module names instead of "multi_process"

File: tests/multi_process_tests.rs [hunk 10]
```diff
@@ -0,0 +1,316 @@
+//! Tests for the multi-process database interface.
+//!
+//! At this step the interface allows only one process to have the database open, so what there is
+//! to test is the directory layout and the lock file that enforces that. The lock is exercised
+//! both from a second handle in this process -- file locks belong to the open file description
+//! rather than the process, so a second handle is excluded exactly as a second process is -- and
+//! from a real child process, which is what the lock is ultimately for.
```
Comment (cberner) [inline tests/multi_process_tests.rs]: No comments about the in-progress work!!!!! How many times do I have to say that. Absolutely under no circumstances are  you allowed to put commentary that reads like a todo list or comments about the task itself. Only comments about the long-term architecture and invariants. Remember that and apply it to all the comments in this whole stack.

**Review by cberner — COMMENTED** [review]
All the comments needs a polish and trimming pass. Also, check that these tests are all appropriate. I don't want to cover every single edge case, just the important parts of the implementation and its contract

**Review by cberner — COMMENTED** [review]
Let's trim the scope of this PR:
* remove begin_write() and the ReadableDatabase impl

The only thing this PR should do is allow create() and open() of a directory structured database. It should also create the metadata file and acquire the required shared lock on it.

**Comment by cberner** [comment]
Applied across the whole stack, all six branches force-pushed:

* **Modules renamed** `multi_process` → `multiprocess` everywhere: the source module, the no_std stub, the test file, the bench file, and the `reject_multiprocess_data_file` helper in 6/6. (The public type names and the `experimental-multiprocess` feature already used the no-underscore spelling and are unchanged.)
* **`StorageBackend` doc paragraph removed.** It was obsolete anyway: under the final design the registry lock serializes header reads against writes, so the backend no longer carries a read/write-atomicity requirement.
* **Comment pass over every branch:** removed everything that narrated the stack or the work ("first step of an incomplete feature", "what a later release needs", "at this step", …) and trimmed the rest down to the architecture and invariants. The module doc that restated the protocol now just points at the design doc section on master.
* **Test pruning:** cut the edge-case tests — symlink markers/planted symlinks, oversized marker, unreadable-directory permissions, `..` paths, and the fine-grained temporary-file hygiene ones — keeping the contract and protocol tests (exclusion, crash recovery of create, snapshot/pinning/invalidation, savepoints, child-process crashes). The guards in the code stay; they just aren't each exercised by a dedicated test any more. Counts per branch: 1/6 has 11 integration tests, the full stack ends at 46 integration + 25 unit.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by cberner** [comment]
Done. This PR is now exactly create() and open() of a directory-structured database: it makes the directory with the write lock and the `metadata` marker, and every open takes the shared lock on `metadata`. `begin_write()`, the `ReadableDatabase` impl, and the doc example are gone from here; they arrive with the protocol in #1387.

The marker moving up from 2/6 turned that PR into the refusal rule that used to sit alongside it — an unmarked directory holding redb's file names without `write.lock` is refused rather than adopted (#1376, retitled). 3/6 onward are unchanged in scope and rebased; PR bodies and the cross-links are updated to match.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/locks.rs [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+    }
+
+    /// Checks that this directory holds a multi-process database, tolerating a missing marker when
+    /// one is being created. A directory holding some other `metadata` file is refused rather than
+    /// taken over, even by `create()`: overwriting it would be a destructive way to report a
+    /// mistyped path. The caller must hold the write lock.
+    fn read_metadata(&self, create: bool) -> Result<(), DatabaseError> {
+        let file = match File::open(self.metadata_file()) {
+            Ok(file) => file,
+            Err(err) if err.kind() == ErrorKind::NotFound => {
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/locks.rs]: Taken, in 915add3. `read_metadata` no longer reads `File::open`'s `NotFound` as an absent marker on its own: the name being occupied at all -- which `symlink_metadata` answers without following the link -- means the directory is not redb's to mark, so a dangling symlink under `metadata` is refused instead of renamed over. `write_metadata_if_missing` asks the same question the same way, so the later `exists()` bypass is gone too.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/locks.rs [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+    ///
+    /// A directory being created is not marked as one of these yet -- the caller does that with
+    /// [`Self::write_metadata_if_missing`], once the database file has turned out to be usable.
+    pub(super) fn open(&self, create: bool) -> Result<Box<dyn StorageBackend>, DatabaseError> {
+        if create {
+            std::fs::create_dir_all(&self.root).map_err(StorageError::Io)?;
+            // Syncing entries inside a directory whose own entry was never flushed loses the lot.
+            // On every create(): finding the directory already there says nothing about whether
+            // anyone flushed it, and the lock lives inside it, so it cannot cover this window
+            sync_parent(&self.root)?;
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/locks.rs]: Taken, in 915add3. `create()` now records the deepest ancestor that already existed before `create_dir_all` runs, and afterwards flushes every directory entry from the database directory's parent up to and including that ancestor, bottom-up. The previous single-parent sync covered only the common case where the parent already existed; a path of freshly made ancestors needed each one flushed, or a crash could drop the whole subtree by losing one entry higher up.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/mod.rs [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            None,
+            self.cache_size,
+            &self.repair_callback,
+        )?;
+        if create {
+            // Last, so that a create() pointed at a directory which turns out not to hold a
+            // database fails without having marked it as one of these
+            dir.write_metadata_if_missing()?;
+        }
+        let metadata_lock = dir.lock_metadata_shared()?;
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/mod.rs]: The ordering is protected by `write.lock` rather than by taking the shared lock earlier, and I have made the convention explicit in `docs/design.md` (in 4/6, which owns the protocol text) instead of restructuring this path.

The opener holds `write.lock` from before it reads the marker until after its shared lock on `metadata` is in place -- in this PR the write lock is held for the handle's lifetime, and in 4/6's multi-writer mode `hold_metadata_lock` runs before the write lock is handed back. So the whole window you describe, `Database::new`'s repair included, sits inside the write lock. An upgrader therefore has to acquire `write.lock` before taking `metadata` exclusively, and the design doc now says so: with that order it cannot start, or swap the inode between validation and locking, while any open is in flight -- and taking the locks the other way round would deadlock against an opener, which is exactly why the order has to be written down rather than left to whoever writes the upgrader.

Locking-then-validating through one handle would still leave the upgrader unable to run during `Database::new` without also holding `write.lock`, so the write-lock convention is needed either way; given that, taking the shared lock last changes nothing observable.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/locks.rs [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+    /// Takes the write lock, which excludes every other process from the database. Taken before
+    /// anything in the directory is read or written, so the holder has the directory to itself.
+    ///
+    /// Only `create()` makes the file, and nothing ever unlinks one: another process may be
+    /// waiting on the lock on that same file, and unlinking would let a third process lock a
+    /// fresh `write.lock` while the second holds the old inode.
+    fn acquire_write_lock(&self, create: bool) -> Result<File, DatabaseError> {
+        let path = self.write_lock_file();
+        let file = if create {
+            open_or_create(&path)
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/locks.rs]: This refusal is the next PR's subject rather than this one's: #1376 adds the pre-lock pass, where an occupied `metadata` name is validated by `read_metadata` before `write.lock` is ever created, so a mistyped `create()` at a foreign directory is refused with the directory left exactly as found -- and the check still repeats under the lock, which stays the authoritative pass. This PR is deliberately trimmed to create/open plus the marker and its shared lock, so the refusal rules, this one included, live one commit up.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/locks.rs [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+
+        Ok(())
+    }
+
+    /// Writes the marker only if the directory does not already carry one. One that is there was
+    /// validated by [`Self::read_metadata`] on the way in, so it is byte-for-byte what would be
+    /// written.
+    pub(super) fn write_metadata_if_missing(&self) -> Result<(), DatabaseError> {
+        if self.metadata_file().exists() {
+            return Ok(());
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/locks.rs]: Taken, in b0af83b. The early return no longer skips the flush: when the marker already exists, `write_metadata_if_missing` still syncs the directory before returning, so a retry that recreated `write.lock` or `data.redb` after a crash lost their entries makes those entries durable rather than reporting success with them still in the air. The marker found there is also validated rather than assumed, which is the same commit's other half.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/locks.rs [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+
+        let write_lock = self.acquire_write_lock(create)?;
+        self.read_metadata(create)?;
+
+        let data = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(create)
+            .truncate(false)
+            .open(self.data_file())
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/locks.rs]: Real, and it lands two PRs up rather than here: #1377 moves initialization under `data.redb.tmp`, renamed into place only once the database has been accepted, so a crash mid-initialization leaves an unambiguous temporary to discard and redo instead of a wedged final name -- with the discard gated on the magic number, so a finished database whose promoting rename was lost is preserved rather than deleted. This PR is trimmed to create/open plus the marker, so the data file here is still initialized in place; the temporary-name scheme arrives with the refusal rules it belongs to.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/multiprocess/locks.rs [hunk 17]
```diff
@@ -0,0 +1,490 @@
...
+        file.lock_shared().map_err(lock_unsupported)?;
+        Ok(file)
+    }
+
+    /// Checks that this directory holds a multi-process database, tolerating a missing marker when
+    /// one is being created. A directory holding some other `metadata` file is refused rather than
+    /// taken over, even by `create()`: overwriting it would be a destructive way to report a
+    /// mistyped path. The caller must hold the write lock.
+    fn read_metadata(&self, create: bool) -> Result<(), DatabaseError> {
+        let file = match File::open(self.metadata_file()) {
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/multiprocess/locks.rs]: This is #1377's subject: `require_regular_file` is introduced there and runs on every name redb opens in place, the marker first among them at the top of `read_metadata`, so a FIFO -- or anything else that is not a regular file -- under `metadata` is refused from a stat rather than opened, and nothing blocks. Within this PR a planted FIFO can indeed make the open wait; the non-regular-file rules all live two PRs up, with the rest of the mistyped-path hardening.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/tree_store/page_store/file_backend/optimized.rs [hunk 18]
```diff
@@ -25,31 +127,64 @@ impl FileBackend {
...
                 warn!(
-                    "File locks not supported on this platform. You must ensure that only a single process opens the database file, at a time"
+                    "Byte-range locks not supported by this filesystem. You must ensure that only a single process opens the database file, at a time"
                 );
 
-                Ok(Self {
-                    file,
-                    lock_supported: false,
-                })
+                false
```
[the remark this answers is not in the document]
Comment (cberner) [reply src/tree_store/page_store/file_backend/optimized.rs:165]: Taken, in 8041632. On an Apple platform where the range lock reports unsupported, the open now falls back to the whole-file lock -- the very filesystem property that makes the fallback necessary is what makes it safe, since without range locks there is nothing for the whole-file lock to conflict with. Two ordinary handles on such a filesystem exclude each other exactly as they did before this change, and a completely unlocked open remains possible only where it always was: on a platform with no file locks at all.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by cberner** [comment]
Closing for a restructure into a two-PR stack: the range-lock addition to the existing open modes alone, and the entire multi-process implementation stacked on top of it, to be split up incrementally as the first merges.

---
_Generated by [Claude Code](https://claude.ai/code)_
